plugins/crypto/openssl/ua_openssl_create_certificate.c

```diff
@@ -0,0 +1,366 @@
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Generating RSA key failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    if(EVP_PKEY_assign_RSA(pkey, rsa) != 1) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Assign RSA key failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+    // rsa will be freed by pkey
+    rsa = NULL;
+
+    // x509v3 has version 2
+    // (https://www.openssl.org/docs/man1.1.0/man3/X509_set_version.html)
+    if(X509_set_version(x509, 2) != 1) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Setting version failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    if(ASN1_INTEGER_set(X509_get_serialNumber(x509), serial) != 1) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Setting serial number failed.");
+        // Only memory errors are possible
+        errRet = UA_STATUSCODE_BADOUTOFMEMORY;
+        goto cleanup;
+    }
+
+    if(X509_gmtime_adj(X509_get_notBefore(x509), 0) == NULL) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Setting 'not before' failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    if(X509_gmtime_adj(X509_get_notAfter(x509), (UA_Int64)60 * 60 * 24 * 356) == NULL) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Setting 'not before' failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    if(X509_set_pubkey(x509, pkey) != 1) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Setting publik key failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    X509_NAME *name = X509_get_subject_name(x509);
+    if(name == NULL) {
+        UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                       "Create Certificate: Getting name failed.");
+        errRet = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    for(UA_UInt32 iSubject = 0; iSubject < lenSubject; ++iSubject) {
+        char field[16];
+        UA_Int32 sep = UA_String_chr(&subject[iSubject], '=');
+        if(sep == -1 || sep == 0 || ((size_t) sep == (subject[iSubject].length - 1)) || sep >= 15) {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
+                           "Create Certificate: Subject must contain one '=' with "
+                           "content before and after.");
+            errRet = UA_STATUSCODE_BADINTERNALERROR;
+            goto cleanup;
+        }
+        memcpy(field, subject[iSubject].data, (size_t) sep);
```

Comment:
You can just replace the '=' with a '\0' in the original string.
Than you don't have to copy to a temporary variable and can just forward the original string with the name.
The replacement can be rolled back later. So that the original string is left identical when returning.

Reply:
This causes a SegFault, when using `UA_STRING_STATIC`.